### ISSUE TITLE
Bug: ConsumerOffsetCommitter goes into failure state after broker downtime

### ISSUE THREAD

**Issue by nioertel** [confluentinc/parallel-consumer#203]
During testing we observed an issue that the `ConsumerOffsetCommitter` apparently tries endlessly to commit offsets without success after a broker downtime.
Test setup:
- Application is started (4 Java processes reading from the same topic which has 24 partitions)
- Application is processing data without issues
- Kafka Broker is killed
- Application has cached messages that were already polled but not processed yet
- Kafka Broker is started again
- Application starts processing again but half of the instances has very low throughput
- In the logs we keep seeing this error:
  ```bash
   [ERROR] 2022-03-01 09:16:37.934 [pc-broker-poll] i.c.p.i.ConsumerOffsetCommitter - Error committing offsets org.apache.kafka.clients.consumer.RetriableCommitFailedException: Offset commit failed with a retriable exception. You should retry committing the latest consumed offsets.
   Caused by: org.apache.kafka.common.errors.DisconnectException
   ```

It seems like the application tries to process messages it still had cached but which were from a partition that was assigned to another application instance and therefor the offset committing doesn't work.

## Related:
- #185 
- #345

**Comment by astubbs** [comment]
Hmmm interesting. Can you post the full stack trace?
And how long has it been left to try to recover after the broker comes back online, before terminating? What was the replication factor of the partition, and can you confirm a new leader was assigned while the broker was offline?

**Comment by astubbs** [comment]
And how many brokers were in this cluster?

**Comment by astubbs** [comment]
Also, while it's continuing to try to commit, do you know if it's still processing new messages or polling more from the brokers?

**Comment by johnbyrnejb** [comment]
Closing Issue

**Comment by astubbs** [comment]
@nioertel - fixed and released in the maintained fork, in [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0). The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```

The short version: in the default asynchronous commit mode, an offset was recorded as committed the moment the request was sent, so a failed or dropped acknowledgement was never retried and the broker's committed offset silently stayed behind. An offset is now recorded as committed only when the broker acknowledges it. The throughput collapse after the broker restart that this issue actually reported did not reproduce against the current code, and the fork has since fixed a large amount in this area, so it is most likely fixed by one of those changes; a recurrence on 0.6.0.0 or later is a new issue there. The full account is on the fork's mirror, astubbs/parallel-consumer#248.
